processing/src/main/java/org/apache/druid/query/aggregation/NullableNumericAggregatorFactory.java

```diff
@@ -85,6 +85,12 @@ public final int getMaxIntermediateSizeWithNulls()
     return getMaxIntermediateSize() + (NullHandling.replaceWithDefault() ? 0 : Byte.BYTES);
   }
 
+  protected boolean canHandleNulls()
+  {
+    return NullHandling.replaceWithDefault();
+  }
```

Comment:
@rohangarg @clintropolis: how do you guys see this api change?
* it could be interpreted as something which is in conflict with the original intention of the interface
  * as it declares in the APIdoc that it will be `null` if nothing is aggregated
* but on the otherhand it somewhat aligns with the original intention as well to give more fine-grained control to the  underlying aggregator on how it would like to handle the `null` values

long story short: the situation here is that a `Long` aggregtor is needed which results in `0` in case of no rows being aggregated - especially in case `useDefaultValueForNull` is turned off

Reply:
This reminds me of an old PR of mine which I totally forgot about until reviewing this PR, #10666, which made a different but somewhat similar change.

In that PR, I was adding a method called `hasNulls` to the nullable numeric aggregator (side note, i probably should resurrect this old PR), to separate processing in SQL compatible mode for columns which have nulls and those which don't, but which should still output a null value if not aggregated.

The changes in that PR would not actually fix the problems here, and I think maybe help me frame what actually bothers me about the changes in this PR, since the changes here are somewhat contradictory to the changes there.

Anyway, I think what bothers me about this change is that it is to handle logic that seems entirely specific to the special case of how COUNT aggregations are handled, and I can't actually see any good use for this otherwise, and because of that I'm not entirely sure it makes sense to add in this way. If we ignore the default value mode (which is now a legacy mode as of #14792), and consider the changes i made in the other PR, then this family of aggregators is for implementing math aggregator functions which may or may not have null values, but should always be null unless aggregated.

COUNT however, is a special case in that it should be 0, even if not aggregated, so I think adding special handling to this family aggregators specifically for the 'merging' aggregator for COUNT seems like the wrong thing to do.

Instead, I think i would be more in favor of just special case implementing the merging/combining aggregator for COUNT, even if it means not sharing some of the same code that we use for SUM, because it is a special case that i can't see having a general use for other types of numeric aggregators, only COUNT.

Reply:
Yes; actually - I wanted a way to declare that this is `NullAware` :D 

Earlier I kinda wanted to take a similar path as you suggested  - just alter the combiner - but that ended up as this change eventually; to my knowledge:

* `CombiningSequence` creates the accumulator and dispatches it thru `Sequence#accumulate`
* as per what I've seen/remember the *first* element is not run thru the combiner as there is nothing to combine it too... 
* there is only a [combine](https://github.com/apache/druid/blob/194a9c9abc1b6edf98c2652687515200c6cbd3e8/processing/src/main/java/org/apache/druid/query/aggregation/CountAggregatorFactory.java#L85) method which has 2 arguments
* or the combining factory...

actually its interesting to consider a corner case (I don't have a testcase for this - but it was covered by existing cases):

* `CombiningSequence` is run on an inputstream which produces only 1 output element - which is produced via the `AggregateCominer` from a number of input
* The `AggregateCombiner` for `CountAggregatorFactory` is `LongSumAggregateCombiner` originally - so it may have returned null ..and thus by giving a sole input value could skip the combine method call

I was considering at some point also to extend the `Accumulator` interface so that the it could communicate the initial value 

Will investigate your suggestion/other alternatives - but iirc when there are 0 elements aggregated those methods are not called/work....that's why I resorted to the current solution.